src/providers/activeLearning/objectDetection.ts

```diff
@@ -0,0 +1,195 @@
+     *     element image, video, or canvas.
+     * @param maxNumBoxes The maximum number of bounding boxes of detected
+     * objects. There can be multiple objects of the same class, but at different
+     * locations. Defaults to 20.
+     *
+     */
+    public async detect(img: tf.Tensor3D|ImageData|HTMLImageElement|HTMLCanvasElement|HTMLVideoElement,
+                        maxNumBoxes: number = 20): Promise<DetectedObject[]> {
+        if (this.model) {
+            return this.infer(img, maxNumBoxes);
+        }
+
+        return [];
+    }
+
+    /**
+     * Infers through the model.
+     *
+     * @param img The image to classify. Can be a tensor or a DOM element image,
+     * video, or canvas.
+     * @param maxNumBoxes The maximum number of bounding boxes of detected
+     * objects. There can be multiple objects of the same class, but at different
+     * locations. Defaults to 20.
+     */
+    private async infer(img: tf.Tensor3D|ImageData|HTMLImageElement|HTMLCanvasElement| HTMLVideoElement,
+                        maxNumBoxes: number): Promise<DetectedObject[]> {
+        const batched = tf.tidy(() => {
+            if (!(img instanceof tf.Tensor)) {
+                img = tf.browser.fromPixels(img);
+            }
+            // Reshape to a single-element batch so we can pass it to executeAsync.
+            return img.expandDims(0);
+        });
+        const height = batched.shape[1];
+        const width = batched.shape[2];
+
+        // model returns two tensors:
+        // 1. box classification score with shape of [1, 1917, 90]
+        // 2. box location with shape of [1, 1917, 1, 4]
+        // where 1917 is the number of box detectors, 90 is the number of classes.
+        // and 4 is the four coordinates of the box.
+        const result = await this.model.executeAsync(batched) as tf.Tensor[];
+
+        const scores = result[0].dataSync() as Float32Array;
+        const boxes = result[1].dataSync() as Float32Array;
+
+        // clean the webgl tensors
+        batched.dispose();
+        tf.dispose(result);
+
+        const [maxScores, classes] = this.calculateMaxScores(scores, result[0].shape[1], result[0].shape[2]);
+
+        const prevBackend = tf.getBackend();
+        // run post process in cpu
+        tf.setBackend("cpu");
+        const indexTensor = tf.tidy(() => {
+            const boxes2 =
+                tf.tensor2d(boxes, [result[1].shape[1], result[1].shape[3]]);
+            return tf.image.nonMaxSuppression(
+                boxes2, maxScores, maxNumBoxes, 0.5, 0.5);
+        });
+
+        const indexes = indexTensor.dataSync() as Float32Array;
+        indexTensor.dispose();
+
+        // restore previous backend
+        tf.setBackend(prevBackend);
+
+        return this.buildDetectedObjects(width, height, boxes, maxScores, indexes, classes);
+    }
+
+    private buildDetectedObjects(
+        width: number, height: number, boxes: Float32Array, scores: number[],
```

Comment:
Lots of parameters here - consider refactoring this into an object.

Reply:
I don't see the point here.  It's a method with 6 parameters.  Is the suggestion about performance, because of the difference of calling by value default type and byref object type, or is this about a better readable code ?

For performance reason I though that copy on write optimization in JS should completely manage this but I'm really eager to learn more here.

For code readability I don't see the point to create an object just to reduce the number of parameters between two private functions but I may be wrong.  It seams to me that wrapping all of this on an object is instead complicating the understanding but once again this is just MHO and eager to learn more.